### ISSUE TITLE
Passing arbitrary attributes through Marathon into Mesos

### ISSUE THREAD

**Issue by clarisca** [SciDAS/NetworkOptimizerService#1]
We need to pass information such as input dataset id and metascheduler ip to Mesos through Marathon.

**Comment by gokkulasudanr92** [comment]
The micro service requires the dataset information and offers to look up from perfSONAR and return a ordered list.
1. The dataset information can be obtained from the TaskInfo Object which is provided to Marathon
2. For offers list, Mesos is required to talk back to the metascheduler from where the request originated. To achieve this we can make use of labels field in the Marathon request object and provide the IP of the metascheduler to talk back (as an key/value pair)

**Comment by mycoy** [comment]
I think there are two steps:
1. Because the labels are only native to Marathon right now, first we need to enable Marathon to send this label information to Mesos. 
2. Enable Mesos to parse this label information. 

So, working on Marathon is unavoidable. We need to learn Scala, Marathon's source code and architecture.

**Comment by gokkulasudanr92** [comment]
@mycoy I did not modify marathon. As I was showing you yesterday, the custom label was automatically reflecting on the mesos frameworks response JSON Object.

**Comment by clarisca** [comment]
@mycoy : Actually we don't need to modify Marathon (see Step 2 in @gokkulasudanr92 's post). This is a built-in capability of Marathon/Mesos.  He was able to create arbitrary new labels and pass them  to Mesos.   We were able to confirm that Mesos does receive the whole object with the new labels. The next step is to go back to Mesos code, extract them and allow the scheduler to act on them (e.g., get label  metascheudler IP and forward offers accordingly).

**Comment by mycoy** [comment]
I don't see any evidence that Marathon can pass this label-information to Mesos.  Please let me know  the exact steps to reproduce this evidence.

**Comment by gokkulasudanr92** [comment]
@mycoy When you create an application using Marathon app POST call use a field ("labels": {"key": "value"}). When you check the http://mesos-ip/frameworks. In the response object list, label Object will display.

**Comment by mycoy** [comment]
This is true. 
But if we want Marathon send this label-information to Mesos-**master** while launching a framework,   basically we want to use Marathon to replace Mesos-**client**,  we still need to modify Marathon to forward the offers to the global scheduler (just like what I've done for mesos-execute).

**Comment by clarisca** [comment]
I don't agree with the statement "...we want to use Marathon to replace Mesos-client". These two components have different functionality and complement each other. We need to modify Mesos code only. In fact, I suspect that is the same part of the code that you already extended. Lets discuss this either tomorrow Friday or next week.

**Comment by mycoy** [comment]
Basically, Marathon is an advanced "mesos-execute", just with more powerful management functions. 
If you create a framework via Marathon, then Marathon's own scheduler will work with Mesos-master,  and this procedure doesn't have anything to do with mesos-execute.

**Comment by clarisca** [comment]
It was a good discussion today. @mycoy @dcvan24.  Thanks.

The idea of pushing the scheduling logic to Marathon has some important disadvantages. More importantly, it breaks the microservice architecture principle (do not overload functionality).  Among the disadvantage of pursuing this direction is that the cost of maintaining backward compatibility.  Every new version of Marathon will trigger an upgrade in our code.

But the whole idea got me thinking that perhaps the best approach is to create a new Mesos framework to handle federation/scheduling. That is, another long running framework (just like Marathon). This approach observes the modularity of the Mesos architecture and hence it does not require modifying Mesos at all. @mycoy 

Since our priority --to start--  is to demonstrate the viability of our approach  I suggest the following:
1- Pass the metascheduler label through _unmodified Marathon_  using the label mechanism that @gokkulasudanr92  suggested.
2- From my conversation with @mycoy I understood that Marathon communicates with the Master only. Hence, lets modify the Master code to write the metasheduler label into the local zookeeper. Zookeper is the  communication channel between Marathon and Mesos anyway. 
This should be a few lines of code --I think.
I have used Zookepeer in the past but only cli. This is a very robust software and well documented. Check: https://github.com/rjenkins/conservator for a c++ client.

3-Modify the client code (already modified anyway) to read the metascheduler label from zookeper.

4-Client uses the metascheduler information to forward the offers to the metascheduler.

There are many solutions to this problem. This is just one of them. For instance,  we could avoid passing the metascheduler label through Marathon and push the logic of selecting a metascheduler and coordinating across clusters to a sideband service. It appears to me that the recommendation above is the simplest. 

What do you all think?
Claris

**Comment by mycoy** [comment]
I look into Mesos master code, to see how to extract that label information.
